sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotOverwriteCollectionElementsBase.cs

```diff
@@ -20,224 +20,100 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Linq;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Diagnostics;
 using SonarAnalyzer.Helpers;
 
 namespace SonarAnalyzer.Rules
 {
-    public abstract class DoNotOverwriteCollectionElementsBase
-        <TInvocationSyntax, TIdentifierSyntax, TStatementSyntax, TMemberAccessSyntax,
-        TThisSyntax, TBaseSyntax, TExpressionSyntax>
-        : SonarDiagnosticAnalyzer
+    public abstract class DoNotOverwriteCollectionElementsBase<TStatementSyntax> : SonarDiagnosticAnalyzer
         where TStatementSyntax : SyntaxNode
-        where TInvocationSyntax : SyntaxNode
-        where TIdentifierSyntax : SyntaxNode
-        where TMemberAccessSyntax : SyntaxNode
-        where TThisSyntax : SyntaxNode
-        where TBaseSyntax : SyntaxNode
-        where TExpressionSyntax : SyntaxNode
     {
         protected const string DiagnosticId = "S4143";
         protected const string MessageFormat = "Verify this is the index/key that was intended; a value has already been set for it.";
 
-        protected abstract SyntaxNode GetExpression(TInvocationSyntax invocation);
+        private static readonly ImmutableArray<KnownType> dictionaryOrCollection = ImmutableArray.Create(
+            KnownType.System_Collections_Generic_IDictionary_TKey_TValue,
+            KnownType.System_Collections_Generic_ICollection_T);
 
-        protected abstract SyntaxNode GetExpression(TMemberAccessSyntax memberAccess);
+        protected abstract DiagnosticDescriptor Rule { get; }
 
-        protected abstract SyntaxToken GetIdentifierToken(TIdentifierSyntax invocation);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(Rule);
 
-        protected abstract SyntaxToken GetIdentifierToken(TMemberAccessSyntax memberAccess);
-
-        protected abstract SyntaxToken GetIdentifierToken(TThisSyntax thisAccess);
-
-        protected abstract SyntaxToken GetIdentifierToken(TBaseSyntax baseAccess);
-
-        protected abstract SyntaxNode GetName(TMemberAccessSyntax memberAccess);
-
-        protected string GetInvokedOnName(TInvocationSyntax invocation)
+        protected void AnalysisAction(SyntaxNodeAnalysisContext context)
         {
-            if (GetExpression(invocation) is TMemberAccessSyntax memberAccessExpression)
+            var statement = (TStatementSyntax)context.Node;
+            var collectionIdentifier = GetCollectionIdentifier(statement);
+            var indexOrKey = GetIndexOrKey(statement);
```

Comment:
Since `GetAssignmentOrInvocation` is language-specific, it should become `abstract` in the base class and the implementations will still be in the VB/C#-specific classes...